neonvm/controllers/vm_controller.go

```diff
@@ -715,41 +729,71 @@ func (r *VMReconciler) doReconcile(ctx context.Context, vm *vmv1.VirtualMachine)
 			// seems already plugged correctly
 			cpuScaled = true
 		}
+		// update status by CPUs used in the VM
+		r.updateVMStatusCPU(ctx, vm, vmRunner, pluggedCPU, cgroupUsage)
 
 		// do hotplug/unplug Memory
 		memSlotsMin := vm.Spec.Guest.MemorySlots.Min
 		targetSlotCount := int(vm.Spec.Guest.MemorySlots.Use - memSlotsMin)
 
-		realSlots, err := QmpSetMemorySlots(ctx, vm, targetSlotCount, r.Recorder)
-		if realSlots < 0 {
-			return err
-		}
-
-		if realSlots != int(targetSlotCount) {
-			log.Info("Couldn't achieve desired memory slot count, will modify .spec.guest.memorySlots.use instead", "details", err)
-			// firstly re-fetch VM
-			if err := r.Get(ctx, types.NamespacedName{Name: vm.Name, Namespace: vm.Namespace}, vm); err != nil {
-				log.Error(err, "Unable to re-fetch VirtualMachine")
+		switch *vm.Status.MemoryProvider {
+		case vmv1.MemoryProviderVirtioMem:
+			targetVirtioMemSize := int64(targetSlotCount) * vm.Spec.Guest.MemorySlotSize.Value()
+			previousTarget, err := QmpSetVirtioMem(vm, targetVirtioMemSize)
+			if err != nil {
 				return err
 			}
-			memorySlotsUseInSpec := vm.Spec.Guest.MemorySlots.Use
-			memoryPluggedSlots := memSlotsMin + int32(realSlots)
-			vm.Spec.Guest.MemorySlots.Use = memoryPluggedSlots
-			if err := r.tryUpdateVM(ctx, vm); err != nil {
-				log.Error(err, "Failed to update .spec.guest.memorySlots.use",
-					"old value", memorySlotsUseInSpec,
-					"new value", memoryPluggedSlots)
+
+			goalTotalSize := resource.NewQuantity(
+				int64(vm.Spec.Guest.MemorySlots.Use)*vm.Spec.Guest.MemorySlotSize.Value(),
+				resource.BinarySI,
+			)
+
+			if previousTarget != targetVirtioMemSize {
+				// We changed the requested size. Make an event for it.
+				reason := "ScaleUp"
+				if targetVirtioMemSize < previousTarget {
+					reason = "ScaleDown"
+				}
+				r.Recorder.Eventf(vm, "Normal", reason, "Set virtio-mem size for %v total memory", goalTotalSize)
+			}
+
+			// Maybe we're already using the amount we want?
+			// Update the status to reflect the current size - and if it matches goalTotalSize, ram
+			// is done.
+			currentTotalSize, err := QmpGetMemorySize(QmpAddr(vm))
```

Comment:
I think this is a redundant check, we could get the current memory size from where it is set. 

But no worries, I'll clean it up in #943

Reply:
I _think_ in testing I actually saw this value change in between QMP queries in the same reconcile cycle. Not totally sure though...